midend/flattenLogMsg.cpp

```diff
@@ -0,0 +1,198 @@
+    newArgument->expression = newStructExpression;
+    newArguments->at(1) = newArgument;
+    newMethodCall->arguments = newArguments;
+    auto* newMethodCallStatement = methodCallStatement->clone();
+    newMethodCallStatement->methodCall = newMethodCall;
+    logMsgReplacament.emplace(methodCallStatement->clone_id, newMethodCallStatement);
+    return newMethodCallStatement;
+}
+
+
+IR::ID FindTypeInLogMsgForReplace::newName() {
+    std::ostringstream ostr;
+    ostr << "f" << index++;
+    return IR::ID(ostr.str());
+}
+
+const IR::Type_StructLike* FindTypeInLogMsgForReplace::generateNewStructType(
+    const IR::Type_StructLike* structType, IR::IndexedVector<IR::NamedExpression> &v) {
+    IR::IndexedVector<IR::StructField> fields;
+    for (auto namedExpr : v) {
+        auto* structField = new IR::StructField(namedExpr->srcInfo, namedExpr->name,
+                                                namedExpr->expression->type);
+        typeMap->setType(structField, structField->type);
+        fields.push_back(structField);
+    }
+    auto* newType = structType->clone();
+    newType->fields = fields;
+    return newType;
+}
+
+TypeLogMsgParams FindTypeInLogMsgForReplace::unfoldStruct(const IR::Expression* expr,
+        std::string strParam, std::string curName) {
+    TypeLogMsgParams result;
+    auto* exprType = typeMap->getType(expr, true);
+    if (!expr->is<IR::StructExpression>()) {
+        if (auto structType = exprType->to<IR::Type_StructLike>()) {
+            result.second += "(";
+            for (auto field : structType->fields) {
+                std::string nm = field->name.name + std::string(":");
+                auto* newMember = new IR::Member(field->type, expr, field->name);
+                if (field->type->is<IR::Type_StructLike>()) {
+                    auto curResult = unfoldStruct(newMember, strParam, field->name.name.c_str());
+                    nm += curResult.second;
+                    result.first.insert(result.first.end(), curResult.first.begin(),
+                                        curResult.first.end());
+                } else {
+                    result.first.push_back(new IR::NamedExpression(expr->srcInfo, newName(),
+                                                                   newMember));
+                    nm += std::string("{}");
+                }
+                if (result.second.length() > 1)
+                    result.second += ",";
+                result.second += nm;
+            }
+            result.second += std::string(")");
+            return result;
+        }
+        result.first.push_back(new IR::NamedExpression(expr->srcInfo, newName(), expr));
+        result.second = curName + std::string(" : {}");
+        return result;
+    }
+    auto* structExpr = expr->to<IR::StructExpression>();
+    std::string strResult;
+    for (auto namedExpr : structExpr->components) {
+        size_t n = strParam.find("{}");
```

Comment:
what happens if you don't find this substring?
should you give a warning?

Reply:
It should add other arguments in StructExpression to this {} like in your example below.